### PR TITLE
ProcessApp Tests

### PR DESCRIPTION
**Pull request by ChristopherDavenport**
Small Testing Suite. Pulling against current release-0.16 and will rebase once the fixes have been incorporated for compliance.

### CODE REVIEW

File: tests/src/test/scala/org/http4s/util/ProcessAppSpec.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+          Process.eval(Task.fail(myError))
+        }
+      }
+      new TestApp().main(Array.empty[String]) should_== (())
+    }
+
+    "Terminate On A Valid Task" in {
+      class TestApp extends ProcessApp {
+        override def main(args: List[String]): Process[Task, Unit] = {
+          Process.emit(Task("Valid Task"))
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/util/ProcessAppSpec.scala]: I think you mean `Process.eval` here.

File: tests/src/test/scala/org/http4s/util/ProcessAppSpec.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      val myError = new Throwable("Bad Initial Process")
+      class TestApp extends ProcessApp {
+        override def main(args: List[String]): Process[Task, Unit] = {
+          Process.fail(myError)
+        }
+      }
+      new TestApp().main(Array.empty[String]) should_== (())
+    }
+
+    "Terminate Server on a Bad Second Value" in {
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/util/ProcessAppSpec.scala]: I'm not sure what this is testing that the first test didn't.  The point is whether the Process fails or not, not that it emitted first.  But test until you're no longer scared, as they say...

File: tests/src/test/scala/org/http4s/util/ProcessAppSpec.scala [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          Process.eval(Task{touched = true; ()})
+        }
+      }
+      new TestApp().main(Array.empty[String]) should_== (())
+    }
+
+    "requestShutdown Shuts Down a Server From A Separate Thread" in {
+      class TestApp extends ProcessApp {
+        override def main(args: List[String]): Process[Task, Unit] = {
+          Process.constant(())
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/util/ProcessAppSpec.scala]: I've written this in a few places but not known what to call it so haven't stuck it in a util.  It runs forever and never emits.  It's what the server config `.serve` does in #1071.  Would be a good fit here, especially if we could name it:

```scala
Process.eval_(Task.async[Nothing] { _ => }))
```

File: tests/src/test/scala/org/http4s/util/ProcessAppSpec.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    "Terminate Server on a Valid Task" in {
+      var touched = false
+      class TestApp extends ProcessApp {
+        override def main(args: List[String]): Process[Task, Unit] = {
+          Process.eval(Task{touched = true; ()})
+        }
+      }
+      new TestApp().main(Array.empty[String]) should_== (())
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/util/ProcessAppSpec.scala]: I was thinking something like this:

```scala
val cleanedUp = async.signalOf(false)
// run forever
Process.eval_(Task.async[Nothing] { _ => }))
  // but observably clean up if stopped
 .onHalt(_ => Process.eval_(cleanedUp.set(true)))
```

And then assert that `cleanedUp` has been set to true.

**Comment by ChristopherDavenport** [comment]
I feel fairly good about the test rig which allows consistent ability to ensure cleanup and that it functions the same in each test.

**Comment by rossabaker** [comment]
I like what you've done with the test rig.  Now the question is why that test fails.

**Comment by ChristopherDavenport** [comment]
Any idea why this would fail as a Task, but succeeds as a Future. Does it perhaps need to be an async task or the like to use as a Task. This passes. 

```scala
val testApp = new TestProcessApp(
        // run forever, emit nothing
        Process.eval_(Task.async[Nothing]{_ => })
      )
      val future = Future(testApp.main(Array.empty[String]))
      testApp.requestShutdown.unsafePerformSync
      Await.result(future.map(_ => testApp.cleanedUp.get.unsafePerformSync), 60.seconds) should_== true
```

**Pull request merged**